src/shared/openssl-util.c

```diff
@@ -1128,6 +1128,97 @@ int string_hashsum(
         return 0;
 }
 #  endif
+
+static int load_key_from_provider(
+                const char *provider,
+                const char *private_key_uri,
+                EVP_PKEY **ret_private_key) {
+
+        assert(provider);
+        assert(private_key_uri);
+        assert(ret_private_key);
+
+#if OPENSSL_VERSION_MAJOR >= 3
+        if (!OSSL_PROVIDER_try_load(/* ctx= */ NULL, provider, /* retain_fallbacks= */ true))
+                return log_openssl_errors("Failed to load OpenSSL provider '%s'", provider);
```

Comment:
i'd prefer if we return a recognizable error in case OSSL_PROVIDER_load() fails, and only fall back to the engine stuff in that case.

i.e. change this to:

```c
if (!OSSL_PROVIDER_try_load(/* ctx= */ NULL, provider, true)) {
        log_openssl_errors("Failed to load OpenSSL provider '%s'", provider);
        return -EADDRNOTAVAIL;
}
```

Reply:
These providers are very new, and completely different implementations from the engine. The problem I'm seeing is with the store load, but someone else might see something different, there's no way to know. For example somebody might have the pkcs11 engine installed, but no pkcs11 provider, as that's pretty much only available in Fedora right now, no other distribution ships it (I have uploaded it to suse/debian/ubuntu but it will take a long time to be approved and then available). So for now it seems far safer to just always try the engine path. In the future, once the providers are more stable and widely available, it can then be restricted.

Reply:
It is true that engines require special code while all providers (including default) are now standardized on the store API.